c/tskit/trees.c

```diff
@@ -2446,79 +2598,107 @@ get_mutation_samples(
+    tsk_size_t r, c, k, s, inner, num_alleles_cumsum;
+    double *result_row;
+    struct stat_matrix_site_indicies idx;
     const tsk_size_t num_samples = self->num_samples;
-    const tsk_size_t max_alleles = self->tables->mutations.num_rows + num_sites;
-    tsk_size_t *num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
-    const tsk_size_t *restrict site_muts_len = self->site_mutations_length;
+    tsk_size_t *num_alleles, *site_offsets;
+    tsk_size_t result_row_len = num_site_cols * result_dim;
+    // tmp needed for reflecting results across axis
+    double *result_row_tmp;
 
     tsk_memset(&allele_samples, 0, sizeof(allele_samples));
 
-    if (num_alleles == NULL) {
+    if (options & TSK_STAT_POLARISED) {
+        polarised = true;
+    }
+
+    get_stat_matrix_site_indices(self->tables->sites.num_rows, num_site_rows, row_sites,
+        num_site_cols, col_sites, &idx);
+
+    num_alleles = tsk_malloc(idx.n_sites * sizeof(*num_alleles));
+    site_offsets = tsk_malloc(idx.n_sites * sizeof(*site_offsets));
+    if (num_alleles == NULL || site_offsets == NULL) {
         ret = TSK_ERR_NO_MEMORY;
         goto out;
     }
 
-    ret = tsk_bit_array_init(&allele_samples, num_samples, max_alleles);
+    num_alleles_cumsum = 0;
+    s = 0;
+    for (s = 0; s < idx.n_sites; s++) {
+        site_id = idx.sites[s];
+        site_offsets[s] = num_alleles_cumsum;
+        num_alleles_cumsum += self->site_mutations_length[site_id] + 1;
+    }
+    ret = tsk_bit_array_init(&allele_samples, num_samples, num_alleles_cumsum);
     if (ret != 0) {
         goto out;
     }
-    ret = get_mutation_samples(self, num_alleles, &allele_samples);
+    ret = get_mutation_samples(self, &idx, num_alleles, &allele_samples);
     if (ret != 0) {
         goto out;
     }
 
-    // Number of pairs w/ replacement (sites)
-    *result_size = (num_sites * (1 + num_sites)) / 2U;
-    *result = tsk_calloc(*result_size * result_dim, sizeof(**result));
+#define COMPUTE_STAT_FOR_MATRIX_IDX(c_id, r_idx, c_idx)                                 \
+    do {                                                                                \
+        tsk_bit_array_get_row(&allele_samples, site_offsets[r_idx], &r_state);          \
+        tsk_bit_array_get_row(&allele_samples, site_offsets[c_idx], &c_state);          \
+        ret = compute_general_two_site_stat_result(&r_state, &c_state,                  \
+            num_alleles[r_idx], num_alleles[c_idx], num_samples, state_dim,             \
+            sample_sets, result_dim, f, f_params, norm_f, polarised,                    \
+            &(result_row[(c_id) *result_dim]));                                         \
+        if (ret != 0) {                                                                 \
+            goto out;                                                                   \
+        }                                                                               \
+    } while (0);
```

Comment:
This macro simplifies the loops below, but I'll admit, it does make debugging a bit more difficult